### ISSUE TITLE
[BUG] SDXL textual inversion produces .safetensors without learned token vector

### ISSUE THREAD

**Issue by Heliox1119** [bmaltais/kohya_ss#3299]
### ⚠️ Describe the bug

When training a textual inversion embedding on SDXL using `sdxl_train_textual_inversion.py`, the training completes successfully and a .safetensors file is generated, but this file contains only the following keys:

`['clip_g', 'clip_l']`

There is no learned vector corresponding to the token defined in `token_string`.

---

### 📁 Dataset structure
```
/home/heliox/IA/Kohya/kohya_ss/dataset/
├── images/
│   └── 40_eliselastnamexl/
│       ├── eliselastnamexl_001.png
│       ├── eliselastnamexl_001.txt
│       ├── eliselastnamexl_002.png
│       ├── eliselastnamexl_002.txt
│       └── ...
├── logs/
│   ├── prompt.txt
│   └── events.out.tfevents.<...>
├── outputs/
│   └── eliselastnamexl_test.safetensors
└── config/
    └── config_textual_inversion.toml
```

---

### 🧪 Reproduction steps

1. Using kohya_ss main (25.2.0)

2. Training is launched with:

```
accelerate launch \
  --num_processes 1 \
  --num_machines 1 \
  --mixed_precision bf16 \
  --dynamo_backend no \
  sd-scripts/sdxl_train_textual_inversion.py \
  --config_file="/home/heliox/IA/Kohya/kohya_ss/config/config_textual_inversion.toml"
```

3. `token_string` is correctly declared in the TOML:

```
bucket_no_upscale = true
bucket_reso_steps = 64
cache_latents = true
caption_extension = ".txt"
clip_skip = 1
dynamo_backend = "no"
enable_bucket = true
epoch = 2
gradient_accumulation_steps = 1
gradient_checkpointing = true
huber_c = 0.1
huber_schedule = "snr"
init_word = "woman"
learning_rate = 1e-4
logging_dir = "/home/heliox/IA/Kohya/kohya_ss/dataset/logs"
loss_type = "l2"
lr_scheduler = "constant_with_warmup"
lr_scheduler_args = []
lr_scheduler_num_cycles = 1
lr_scheduler_power = 1
lr_warmup_steps = 500
max_bucket_reso = 1536
max_data_loader_n_workers = 0
max_timestep = 1000
max_token_length = 75
max_train_steps = 200
mem_eff_attn = true
min_bucket_reso = 256
mixed_precision = "bf16"
multires_noise_discount = 0.3
no_half_vae = true
noise_offset_type = "Original"
num_vectors_per_token = 1
optimizer_args = []
optimizer_type = "AdamW"
output_dir = "/home/heliox/IA/Kohya/kohya_ss/dataset/outputs"
output_name = "eliselastnamexl_test"
pretrained_model_name_or_path = "stabilityai/stable-diffusion-xl-base-1.0"
prior_loss_weight = 0
resolution = "1024,1536"
sample_prompts = "/home/heliox/IA/Kohya/kohya_ss/dataset/logs/prompt.txt"
sample_sampler = "euler_a"
save_every_n_epochs = 1
save_model_as = "safetensors"
save_precision = "bf16"
sdpa = true
token_string = "eliselastnamexl"
train_batch_size = 1
train_data_dir = "/home/heliox/IA/Kohya/kohya_ss/dataset/images"
```

4. Training logs clearly show that the token is registered and training runs:

```
tokens are added for tokenizer 1: [49408]
tokens are added for tokenizer 2: [49408]
create embeddings for 1 tokens, for eliselastnamexl
```

5. Loss decreases normally during training (e.g., ~0.13 at step 127)

6. Final `.safetensors` is generated successfully, but when inspecting:

```
from safetensors.torch import load_file
print(load_file("eliselastnamexl_test.safetensors").keys())
```

It returns:

`['clip_g', 'clip_l']`

→ The token vector is missing.

### 📄 Logs

📎 Attached:
●  [loss_export.csv](https://github.com/user-attachments/files/20845958/loss_export.csv) (exported from TensorBoard)
→ Confirms that training ran correctly and produced coherent loss values step-by-step.

---

### ✅ Expected behavior

The final `.safetensors` (and intermediate `.safetensors` files if saved) should include the learned vector associated with the token specified in `token_string`.

Note:
Every `.safetensors` file (final or intermediate, regardless of training step) has the exact same size: 4.7 KB.
This strongly suggests that no learned vector is stored inside, regardless of `save_every_n_epochs` or `max_train_steps`.

---

### 🖥️ Environment

`OS`: Ubuntu 22.04.5

`GPU`: AMD RX 7900 XTX (ROCm 6.3.2)

`kohya_ss`: main (25.2.0)

`accelerate`: 1.8.1

`torch`: compiled for ROCm (2.7.1+rocm6.3)

---

### 🔁 Additional context

This occurs with SDXL textual inversion training, using the current main branch.
The training logs are clean, and all steps appear to execute normally, but the learned vector is not present in any output files.

I encountered the exact same issue (empty `.safetensors` file after training) using the [hqnicolas/bmaltaisKohya_ssROCm](https://github.com/hqnicolas/bmaltaisKohya_ssROCm) fork, which is dedicated to AMD ROCm.

It was precisely because of this bug that I chose to abandon that fork and switch back to the official main branch of kohya_ss, which also supports ROCm natively.

So the problem seems to be tied directly to the training script itself (`sdxl_train_textual_inversion.py`) or the way embeddings are written to `.safetensors`, regardless of the backend.

🙏 Please investigate and confirm if there's a bug in the saving logic.

**Comment by bmaltais** [comment]
Hi, this is probably an issue with kohya sd-scripts. I don’t think he has given much attention to the TI portion in a long time. You should report this bug on his GitHub repository so he can fix it. Once fixed I will update the version of sd-scripts used by the gui.

**Comment by Heliox1119** [comment]
Thanks for your quick response and advice!
I noticed that the sdxl_train_textual_inversion.py script on the original kohya-ss/sd-scripts repository might be more up-to-date than the one currently used in your GUI: https://github.com/kohya-ss/sd-scripts/blob/main/sdxl_train_textual_inversion.py.

**Comment by Heliox1119** [comment]
@bmaltais 

✅ Update: I can now confirm that after switching to the latest sdxl_train_textual_inversion.py script from the official [kohya-ss/sd-scripts](https://github.com/kohya-ss/sd-scripts) repository, textual inversion training works as expected. The `clip_l` and `clip_g` keys are properly saved into the `.safetensors` file, and the learned embedding behaves correctly at inference.

Thanks again for your guidance — your intuition was right. The bug seems to be limited to the older TI script bundled in the current GUI build.
